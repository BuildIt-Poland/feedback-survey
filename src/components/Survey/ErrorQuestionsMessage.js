import React from 'react';
import { string } from 'prop-types';
import { Box, Heading } from 'grommet';

import { colorRed } from '../../styles/variables';

const ErrorQuestionsMessage = ({ message }) => (
  <Box align="center" justify="center" pad="medium" height="90vh">
    <Heading level="4" textAlign="center" color={colorRed}>
      {message}
    </Heading>
  </Box>
);

ErrorQuestionsMessage.propTypes = {
  message: string.isRequired
};

export default ErrorQuestionsMessage;
