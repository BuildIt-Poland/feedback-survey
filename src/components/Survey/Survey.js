import React from 'react';
import { Redirect } from 'react-router-dom';

import { Form, Box, Heading } from 'grommet';

import { fetchQuestions, saveSurvey } from '../../services/api';
import SurveyField from '../SurveyField/SurveyField';
import SurveyButton from '../Button';

const messages = {
  required: 'This field is required'
};

class Survey extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: props.isLoading || true,
      data: { questions: [], answerTypes: {} },
      submitted: false,
      success: false,
      error: null
    };
  }

  componentDidMount() {
    if (this.state.isLoading) {
      fetchQuestions().then(res => {
        this.setState({
          data: res.data,
          isLoading: false
        });
      });
    }
  }

  renderSurvey() {
    const { submitted, success, error, data, isLoading } = this.state;
    if (submitted) {
      return (
        <Redirect
          to={{
            pathname: '/final',
            state: { success, error }
          }}
        />
      );
    } else {
      if (isLoading) {
        return (
          <Box animation="pulse" align="center" justify="center" pad="medium" height="90vh">
            <Heading level="4">Loading...</Heading>
          </Box>
        );
      } else {
        return (
          <Box align="center" justify="center" pad="medium">
            <Form onSubmit={this.handleSubmit} messages={messages}>
              {this.renderFormFields(data)}
              <Box align="center">
                <SurveyButton type="submit" label="submit" margin="xlarge" data-testid="survey-submit-button" />
              </Box>
            </Form>
          </Box>
        );
      }
    }
  }

  renderFormFields(data) {
    return data.questions.map((field, index) => (
      <SurveyField key={index} field={field} answerTypes={data.answerTypes} />
    ));
  }

  handleSuccess = () => {
    this.setState({
      submitted: true,
      success: true
    });
  };

  handleError = error => {
    this.setState({
      submitted: true,
      error: error.message
    });
  };

  handleSubmit = event => {
    const data = {
      surveyId: this.props.match.params.surveyId || '',
      answers: event.value
    };
    saveSurvey(data).then(res => this.handleSuccess(), err => this.handleError(err));
  };

  render() {
    return this.renderSurvey();
  }
}

export default Survey;
