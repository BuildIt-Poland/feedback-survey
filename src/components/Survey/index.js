import { withRouter } from 'react-router-dom';

import Survey from './Survey';

export default withRouter(Survey);
