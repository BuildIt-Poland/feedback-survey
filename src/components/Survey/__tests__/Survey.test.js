import React from 'react';
import { create } from 'react-test-renderer';

import Survey from '../Survey';

describe('COMPONENT - Survey', () => {
  it('renders info message if there are no questions ', () => {
    const data = {
      questions: [],
      answerTypes: []
    };
    const component = create(<Survey data={data} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const data = {
      questions: [
        {
          id: '4',
          type: 'select',
          content: 'What kind of feedback are you providing?',
          required: true,
          name: 'kind-of-feedback',
          answerType: 'kindOfFeedback'
        }
      ],
      answerTypes: [
        {
          type: 'kindOfFeedback',
          values: ['monthly', 'quaterly', 'annual']
        }
      ]
    };
    const component = create(<Survey data={data} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
