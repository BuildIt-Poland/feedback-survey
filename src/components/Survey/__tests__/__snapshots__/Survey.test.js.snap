// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - Survey renders Loading component if data loading 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 hUosMD"
>
  <h4
    className="StyledHeading-sc-1rdh4aw-0 hlZfma"
  >
    Loading...
  </h4>
</div>
`;

exports[`COMPONENT - Survey renders correctly with data 1`] = `
<main
  className="sc-dnqmqq iACYhp"
>
  <section
    className="sc-htoDjs ggzxba"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="StyledGrommet-sc-19lkkz7-0 eUfLDp"
      >
        <div
          className="sc-EHOje cpubXY"
        >
          <div
            className="sc-ifAKCX dDQvBh FormField__FormFieldBox-m9hood-0 igkoAa StyledBox-sc-13pk1d4-0 CAsMy"
          >
            <label
              className="StyledText-sc-1sadyjn-0 bMXxti"
            >
              What kind of feedback are you providing?
            </label>
            <div
              className="StyledBox-sc-13pk1d4-0 gZSidW"
            >
              <button
                aria-label="Open Drop"
                className="Select__StyledSelectDropButton-sc-17idtfo-1 gQjIFS StyledButton-sc-323bzc-0 jfbKsv"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                type="button"
              >
                <div
                  className="StyledBox-sc-13pk1d4-0 erazuK"
                >
                  <div
                    className="StyledBox-sc-13pk1d4-0 XINUX"
                  >
                    <div
                      className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
                    >
                      <input
                        autoComplete="off"
                        className="Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 cyWwXE"
                        name="kind-of-feedback"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        readOnly={true}
                        tabIndex="-1"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    className="StyledBox-sc-13pk1d4-0 eLQtUu"
                    style={
                      Object {
                        "minWidth": "auto",
                      }
                    }
                  >
                    <svg
                      aria-label="FormDown"
                      className="StyledIcon-ofa7kd-0 kKIhud"
                      viewBox="0 0 24 24"
                    >
                      <polyline
                        fill="none"
                        points="18 9 12 15 6 9"
                        stroke="#000"
                        strokeWidth="2"
                      />
                    </svg>
                  </div>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="StyledBox-sc-13pk1d4-0 dNgQiZ"
      >
        <button
          className="sc-bZQynM iuADUF StyledButton-sc-323bzc-0 bmExQD"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          size="large"
          type="submit"
        >
          submit
        </button>
      </div>
    </form>
  </section>
</main>
`;

exports[`COMPONENT - Survey renders error message 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 ePVEiU"
>
  <h4
    className="StyledHeading-sc-1rdh4aw-0 btqNMh"
  >
    error message
  </h4>
</div>
`;

exports[`COMPONENT - Survey renders info message if there are no questions  1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 ePVEiU"
>
  <h4
    className="StyledHeading-sc-1rdh4aw-0 btqNMh"
  >
    Oops, something went wrong, there is no questions.
  </h4>
</div>
`;
